core/src/main/java/org/infinispan/util/concurrent/BlockingManager.java

```diff
@@ -174,10 +174,13 @@
    <V> CompletionStage<V> continueOnNonBlockingThread(CompletionStage<V> delay, Object traceId);
 
    /**
-    * Provided a publisher that is known to block when subscribed to, this ensures that the publisher is subscribed
-    * on the blocking executor and any values published are observed on a non blocking thread. Note that if a
-    * blocking thread subscribes to the publisher these additional threads are not used and the entire Publisher
-    * is subscribed and observed on the invoking thread.
+    * Provided a publisher that is known to block when subscribed to. Thus if the thread that subscribes in a non
```

Comment:
They should certainly care. The point of this method is to only do the subscribe and request part in a blocking thread but any published values should be on the non blocking thread. We don't want to expose the blocking thread to the returned Publisher.

Reply:
I would argue they should not care, because the javadoc later says "If, however, the subscribing thread is a blocking thread no threading changes will be done, which means the publisher will be subscribed to on the invoking thread. In this case values have no guarantee as to which thread they are observed on, dependent solely on how the Publisher publishes them."

`NonBlockingStore.publishKeys()` doesn't say anything about the observing thread/scheduler either. It only says says subscribing should not block the caller.

If the callers do expect values to be observed on a non-blocking thread, then I think we should either call `observeOn()` when the caller is blocking as well, or assert that the caller thread is non-blocking.

Reply:
> `NonBlockingStore.publishKeys()` doesn't say anything about the observing thread/scheduler either. It only says says subscribing should not block the caller.

Yes, because the implementors shouldn't care about the thread it is done as long as it doesn't block. The SPI is more for the implementors not the invokers in my book.

Reply:
I agree the SPI javadoc is for store implementors. But if the javadoc doesn't explicitly tell them that the publisher they return must emit its values on a non-blocking thread, they may not see any reason to use `BlockingManager.blockingPublisher()`, and emit the values on a blocking thread instead. Or they may call `BlockingManager.blockingPublisher()`, but from a blocking thread, causing the values to be emitted on the same blocking thread.

That's why I think it's important for the SPI to be explicit about what kind of thread it expects the result publisher to emit its values on, not just to recommend using `BlockingManager.blockingPublisher()`.

OTOH if the thread where the values are emitted doesn't matter, then `BlockingManager.blockingPublisher()` shouldn't make any promises about it either, because stores are its only users (at least for now).

Reply:
> I agree the SPI javadoc is for store implementors. But if the javadoc doesn't explicitly tell them that the publisher they return must emit its values on a non-blocking thread, they may not see any reason to use `BlockingManager.blockingPublisher()`, and emit the values on a blocking thread instead. Or they may call `BlockingManager.blockingPublisher()`, but from a blocking thread, causing the values to be emitted on the same blocking thread.
> 

Implementers can publish on a blocking thread if they want, as long as the invoking thread is a blocking thread. We could be more specific and say don't block the invoking thread if it is invoked in a non blocking thread, but I stayed away from that as it puts a lot of burden on the implementer to understand that. That is why I insulated all of that to the `BlockingManager` in the first place and the user just uses that if the operation can block and it will transparently handle which thread to subscribe/publish based on the invoking thread and which part of it is blocking.

> That's why I think it's important for the SPI to be explicit about what kind of thread it expects the result publisher to emit its values on, not just to recommend using `BlockingManager.blockingPublisher()`.

I disagree. If the operation can block they should use `BlockingManager` to isolate the parts that can block and that code will do the right thing.

> OTOH if the thread where the values are emitted doesn't matter, then `BlockingManager.blockingPublisher()` shouldn't make any promises about it either, because stores are its only users (at least for now).

It does matter, it matters based on the invoking thread, which is what `BlockingManager` handles for the user. And I foresee other uses of `blockingPublisher` as that becomes more mainstream in the code.

Reply:
> Implementers can publish on a blocking thread if they want, as long as the invoking thread is a blocking thread.

Except this requirement is not documented, so implementors will ignore it.

E.g. in `RemoteStore`, it seems to be ok for `publishEntries()` to emit the entries on a remote cache thread, although that means there could be more CPU-bound threads in use than the user set in their configuration (or that we picked based on the number of CPUs).

But say someone is writing a store like `RemoteStore`, only `publishEntries()` also has to log the entries to a local file before it emits them. According to the javadoc, they could submit the logging to the blocking thread pool and emit the entry on the same thread. `remoteCache.publishEntries()` does not block, so they don't see any reason to use `BlockingManager.blockingPublisher()`.